packages/typescript/src/index.ts

```diff
@@ -82,22 +81,24 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
       if (!filter(id)) return null;
 
       const output = findTypescriptOutput(ts, parsedOptions, id, emittedFiles);
-      output.declarations.forEach((declaration) => declarationFiles.add(declaration));
 
       return output.code ? (output as SourceDescription) : null;
     },
 
     generateBundle(outputOptions) {
-      for (const id of declarationFiles) {
-        const code = emittedFiles.get(id);
-        if (code) {
+      parsedOptions.fileNames.forEach(fileName => {
+        const output = findTypescriptOutput(ts, parsedOptions, fileName, emittedFiles);
+        output.declarations.forEach((id) => {
```

Comment:
We should probably split `findTypescriptOutput` into separate functions later on. I'm fine with merging this and making changes later if you like.